sdk/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java

```diff
@@ -0,0 +1,99 @@
+package io.opentelemetry.sdk.common.export;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * The implementation of Export operations are often asynchronous in nature, hence the need to
+ * convey a result at a later time. CompletableResultCode facilitates this.
+ *
+ * <p>This class models JDK 8's CompletableFuture to afford migration should Open Telemetry's SDK
+ * select JDK 8 or greater as a baseline, and also to offer familiarity to developers.
+ */
+public class CompletableResultCode {
+  /** A convenience for declaring success */
+  public static final CompletableResultCode SUCCESS = new CompletableResultCode().success();
+
+  /** A convenience for declaring failure */
+  public static final CompletableResultCode FAILURE = new CompletableResultCode().failure();
+
+  public CompletableResultCode() {}
+
+  private volatile Boolean succeeded = null;
+  private final ArrayList<Runnable> actions = new ArrayList<>();
+  private final Lock state = new ReentrantLock();
```

Comment:
Since we have an object for this lock now, is there any need for it to be a `ReentrantLock` as opposed to an `Object` which we use `synchronized` on? IIRC, the latter is a bit faster with a bit less syntax (though maybe not in Java 15 where they're finally removing biased locking)

Reply:
What happens when the object is null?

Also, do you have a ref re. `synchronized` outperforming `ReentrantLock`. I'd imagine that the former is implemented using the latter via pthreads, at least on Unix.

Reply:
It wouldn't ever be `null` since you'd declare `private final Object lock = new Object()`. We use the pattern in the span itself too, as do many JDK classes, so it'd be a bit more consistent too.

https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsReadableSpan.java#L73

Not sure about all platforms, but on the main ones Java doesn't use OS for locks, it's all in the JVM using CAS, including crazy stuff that only applies for `synchronized` AFAIK (though it's going away)

https://wiki.openjdk.java.net/display/HotSpot/Synchronization

Here I'd err on consistency with the span and a couple lines of code less if there's no specific need for the `ReentrantLock`.

Reply:
Sorry, I misunderstood which object you were referring to.

We actually do need the `ReentrantLock` in case an `andThen` action calls upon other parts of the API, right?

Reply:
Actually, your suggestion appears to support reentrancy also (I didn't know! https://docs.oracle.com/javase/tutorial/essential/concurrency/locksync.html). I shall change to be consistent with other parts of the code.

Reply:
Changes pushed